runtime/src/master.rs

```diff
@@ -0,0 +1,695 @@
+//! `decl_module!` macro. Let's call that enum `module::Call`. Each `module::Call` is a reified
+//! invocation of one of the modules methods. The `module::Call` for this:
+//!
+//! ```
+//! # mod a {
+//! # use frame_support::{decl_module, dispatch::DispatchResult};
+//! # type Foo = ();
+//! # type Bar = ();
+//! # trait Trait: system::Trait {}
+//! decl_module! {
+//!     pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+//!         #[weight = 100_000]
+//!         pub fn frob(origin, foo: Foo) -> DispatchResult { Ok(()) }
+//!         #[weight = 100_000]
+//!         pub fn unfrob(origin, foo: Foo, bar: Bar) -> DispatchResult { Ok(()) }
+//!     }
+//! }
+//! # }
+//! ```
+//!
+//! looks something like this:
+//!
+//! ```
+//! # type Foo = ();
+//! # type Bar = ();
+//! enum Call {
+//!     frob(Foo),
+//!     unfrob(Foo, Bar),
+//! }
+//! ```
+//!
+//! The `construct_runtime!` macro assembles the calls from all the included modules into a
+//! single "super call". The name of this enum is also "Call", but let's refer to it as
+//! `runtime::Call`. A `runtime::Call` looks something like this:
+//!
+//! ```
+//! # mod module1 { pub type Call = (); }
+//! # mod module2 { pub type Call = (); }
+//! # mod module3 { pub type Call = (); }
+//! enum Call {
+//!    Module1(module1::Call),
+//!    Module2(module2::Call),
+//!    Module3(module3::Call),
+//! }
+//! ```
+//!
+//! This module allows members of the group called Master to "vote" on a `runtime::Call` (a
+//! proposal) using cryptographic signatures. The votes, along with the proposed `runtime::Call`
+//! are submitted in a single transaction. If enough valid votes endorse the proposal, the proposal
+//! is run as root. If the running the proposal as root succeeds, a new round of voting is started.
+//!
+//! Each member of Master is idenitified by their dock DID.
+//!
+//! This module implement partial replay protection to prevent unauthorized resubmission of votes
+//! from previous rounds.
+
+// REVIEWER: the Membership type is capable of soundly representing
+//           a superset of revoke::Policy::OneOf. Should we unify the
+//           types? It would make the revoke module more flexible and
+//           would be code re-use.
+// ```
+// enum Policy {
+//     Vote(Membership),
+// }
+// ```
+
+use crate::{
+    did::{Did, DidSignature},
+    StateChange,
+};
+use alloc::{
+    boxed::Box,
+    collections::{BTreeMap, BTreeSet},
+    vec::Vec,
+};
+use codec::{Decode, Encode};
+use core::default::Default;
+use frame_support::{
+    decl_error, decl_event, decl_module, decl_storage,
+    dispatch::{DispatchResult, Dispatchable},
+    ensure,
+    weights::GetDispatchInfo,
+    Parameter,
+};
+use system::{ensure_root, ensure_signed};
+
+#[derive(Encode, Decode, Clone, PartialEq, Debug)]
+pub struct Vote {
+    /// The serialized Call to be run as root.
+    // REVIEWER: The type of this call is erased becuase vote needs to go in StateChange
```

Comment:
Nope.

Reply:
Can you check?

Reply:
Testing master voting programmatically will require setting up an rpc client which I have not done.
Since the type of "Call" is erased and replaced with a `Vec<u8>` the polkadot web gui won't know how to let the user construct it interactively.

What would you like me to check?

Reply:
If this type erasure is a problem, one option is to split the statechange enum up and add unique tags for each type of state change.

```rust
trait StateChange: Encode {
    const PREFIX: [u8; 32];
}

impl StateChange for did::KeyUpdate {
    const PREFIX: [u8; 32] = hex!("0f07937ee218041d2c7006e1a58087e0de8de695aef5f1adfcad01c3e623a10a");
}

#[test]
fn key_update_prefix() {
    assert_eq!(
        <did::KeyUpdate as StateChange>::PREFIX,
        hash("Authorize a single update to a DID on the dock blockchain.")
    );
}
```

Reply:
The objective is to know that it works and if you can check that even 1 of the sudo calls from PoA can be made through Master. Checking with polkadot-js would be easy..

Reply:
Understood and agreed. Added a jira issue https://dock-team.atlassian.net/browse/DCK-292 .